LSP fixes

mlocati opened the pull request:
We have still a few LSP collisions (see #2714 )... What about fixing them so that concrete5 can run on the upcoming PHP 7? :wink:

Let's start with `Attribute\Key::getList`: it conlflicts with the `getList` methods implemented in the child classes (eg  `UserKey`).

With this pull request I'm suggesting to rename `Attribute\Key::getList` to `Attribute\Key::get` (someone can think of better name?), while keeping a backward compatibility by using `__callStatic`.

mlocati commented:
Nope, `Attribute\Key::get` is already in use by one of the class child classes (`FileKey::get`).
Any suggestion for a new name of `Attribute\Key::getList`?

joe-meyer commented:
what about `getCategoryAttributes` or `getByCategory` ?

mlocati commented:
> what about getCategoryAttributes or getByCategory ?

@aembler ?

aembler commented:
How about Key::getAttributeKeyList()

KorvinSzanto commented:
Isn't there a better place that we can move this functionality to observe SRP [1]? We can deprecate the old metho and provide the functionality elsewhere.

[1] https://en.wikipedia.org/wiki/Single_responsibility_principle

mlocati commented:
The problem here is that the old method must not be defined (since it has a signature that differ from  the ones of the child classes): it's not enough not to call it. That's the reason why I removed it and added a __callStatic.
About the SRP, I think that this method already respects it: it's the base class for the attribute keys, and this method handles all the types of the  attribute keys...

KorvinSzanto commented:
@mlocati This class should be a value object, it shouldn't be doing anything other than being an attribute key according to SRP. I'd contend that getting a list of keys in an attribute key category has no business being on the attribute key class, I'd entertain the idea of it being on the category class, but I still don't think that'd satisfy the spirit of SRP.

To be clear, I'm suggesting we shim for backwards compatibility in the way that @mlocati described and  potentially add a class docblock that marks the shimmed method deprecated while referring people to the new method.

mlocati commented:
> @mlocati This class should be a value object, it shouldn't be doing anything other than being an attribute key according to SRP.

Well, I think that such a strict approach to SRP is somehow quite counterintuitive. Adopting it would lead to completly avoid static methods for value objects.
For instance, instead of
```php
if ($myCar->sellAtAReasonablePrice()) {
    $myCar = Car::buyFromDealer('Alfa Romeo Giulietta');
}
```
We should have
```php
if ($myCar->sellAtAReasonablePrice()) {
    $myCar = CarDealer::buyCar('Alfa Romeo Giulietta');
}
```
(yes, I'm probably going to buy a new car :wink:)
BTW, I think that this approach is quite dispersive, since we may have a `CarDealer::buyCar` as well as `PrivateCitizen::buyCar`, `Lottery::winCar` or even `CarThief::recycleCar` :wink: ).
And in those cases too we won't respect the strict SRP approach, since `CarDealer`, ... shouldn't have static methods.
So, we should switch to a service class, but doing so we would have two classes (model + service) for every kind of object... What a mess! I really do find more easy to understand, to remember and to use a single class to work with cars. Wanna buy a car? Type `Car::` and let your IDE suggest you the available static methods...

mlocati commented:
Ok, much better [now](https://travis-ci.org/concrete5/concrete5/jobs/87626716) than [before](https://travis-ci.org/concrete5/concrete5/jobs/87359839).
We have still these items remaining:
- `Concrete\Core\Gathering\Item\Page::add` vs `Concrete\Core\Gathering\Item\Item::add`
- `Concrete\Core\File\Set\SavedSearch::add` vs `Concrete\Core\File\Set\Set::add`
- `Concrete\Core\Gathering\Item\FlickrFeed::add` vs `Concrete\Core\Gathering\Item\Item::add`
- `Concrete\Core\Gathering\Item\Page::add` vs `Concrete\Core\Gathering\Item\Item::add`
- `Concrete\Core\Gathering\Item\RssFeed::add` vs `Concrete\Core\Gathering\Item\Item::add`
- `Concrete\Core\Gathering\Item\Twitter::add` vs `Concrete\Core\Gathering\Item\Item::add`
- `Concrete\Core\Attribute\PendingType::getList` vs `Concrete\Core\Attribute\Type::getList`
- `Concrete\Core\Workflow\Progress\PageProgress::add` vs `Concrete\Core\Workflow\Progress\Progress::add`

mlocati commented:
> Concrete\Core\Attribute\PendingType::getList vs Concrete\Core\Attribute\Type::getList

Fixed in https://github.com/concrete5/concrete5/commit/4e0bd642b3ad01aa544f3e74a6f3fe0945e81379

mlocati commented:
@aembler @KorvinSzanto   If this PR is going in the right direction for you I can go on and fix the remaining issues I [mentioned above](https://github.com/concrete5/concrete5/pull/3058#issuecomment-151425232)...

aembler commented:
It looks like Key::getList() has changed into Key::get() – I think that might make more sense as Key::getAttributeKeyList(), especially since Type::getList() has been changed into Type::getAttributeTypeList(). Maybe you meant to change Key::getList instead of Type::getList() ?

Regarding the SRP talk – that's a good goal to work toward. But let's not make perfect the enemy of the good right now. This isn't the place to make wide-ranging decisions about refactoring. This is the place to get fixes into the core that will allow concrete5 to work on PHP 7 Right Now™. Thanks @mlocati

mlocati commented:
> It looks like Key::getList() has changed into Key::get() – I think that might make more sense as Key::getAttributeKeyList()

`Key::get` was introduced in the [3rd commit](https://github.com/mlocati/concrete5/commit/25ac3fcc86389409382fd334cb4506414db69a04) of this PR.
In the [5th commit](https://github.com/mlocati/concrete5/commit/bacba841762ef551ad5eddc0b161db191ef64341) I renamed the method to `Key::getAttributeKeyList` (as per [your suggestion](https://github.com/concrete5/concrete5/pull/3058#issuecomment-151283184) and because it clashed with `FileKey::get`)


> especially since Type::getList() has been changed into Type::getAttributeTypeList()

That was done in the [10th commit](https://github.com/mlocati/concrete5/commit/4e0bd642b3ad01aa544f3e74a6f3fe0945e81379), for which I adopted the same naming that you soggested for Key.

> Maybe you meant to change Key::getList instead of Type::getList() ?

We need to change both, since both `Key` and `Type` have child classes that implements a static `getList` method with a different signature (ie different parameters).

aembler commented:
My fault. I was reading the commits up from the bottom. Yes, that all makes sense then.

KorvinSzanto commented:
@mlocati 

> @mlocati This class should be a value object, it shouldn't be doing anything other than being an attribute key according to SRP.

> Well, I think that such a strict approach to SRP is somehow quite counterintuitive. Adopting it would lead to completly avoid static methods for value objects.

SRP doesn't speak to static vs non-static, just that classes have only a single reason to change

>(yes, I'm probably going to buy a new car :wink:)
>BTW, I think that this approach is quite dispersive, since we may have a `CarDealer::buyCar` as well as `PrivateCitizen::buyCar`, `Lottery::winCar` or even `CarThief::recycleCar` :wink: ).
> And in those cases too we won't respect the strict SRP approach, since `CarDealer`, ... shouldn't have static methods.

I think you're looking at this wrong. In real life, I do not buy tell my car what dealership to purchase itself from, in reality the car has almost nothing to do with the purchase.

    <?php
    $dealership->buyCar($car)
    // vs
    $lottery->winCar($car);
    // vs 
    $thief->stealCar($car);

This is a great example and really highlights the confusion that not following SRP causes. Notice that in this example, $car has nothing to do with the dealership logic or the lottery logic or the thief logic. If I want to change how a dealership sells a car, why should I have to change the Car class??? This is why we have SRP, we do not have SRP to somehow eliminate static methods or otherwise make things completely disperate. We have SRP to separate responsibilities.

> So, we should switch to a service class, but doing so we would have two classes (model + service) for every kind of object... What a mess! I really do find more easy to understand, to remember and to use a single class to work with cars. Wanna buy a car? Type `Car::` and let your IDE suggest you the available static methods...

You're missing the point though, you forget that there's more context to buying a car than just the car.

```
Car::buyCar($dealership)
```
Now my Car class is aware of the dealership object and has to have logic to enable purchasing from a dealership

```
Car::buyCar($person)
```
Okay, now my car can handle purchasing itself from a person, now Car is tightly coupled to Person too and contains private purchase logic

If I want to test my car class, I now have to test not only the car's value but I have to test the complete flow of purchasing a car. Further if two separate people are working on my application at once, it's not unlikely that one person would want to change how a car is purchased while another team will want to change the car's value, this will lead to conflicts.

We have SOLID for a reason @mlocati I think you'd really do yourself a favor by [reading into SRP](http://code.tutsplus.com/tutorials/solid-part-1-the-single-responsibility-principle--net-36074) and SOLID as a whole.

Like I said though, I'd be willing to entertain the idea of moving this logic to the CATEGORY class, where at least it makes a little bit of sense rather than implementing a complex `KeyListFactory` or something like that.

joe-meyer commented:
I'm late to the party but I do think we're doing these backwards. If we think about these in the context of entity relations for example, `AttributeType::getAttributeTypeList('file')` is going to have to look up the Category, then get the related types from that category, why not just have a `AttributeCategory->getTypes` or `AttributeCategory->getAttributeKeys()`? It's still not perfect SRP but it does future proof us a bit in preperation for doctrine entities. 

To elaborate on the car example I feel like right now we're essentially doing a `Car::getListOfCarsForSale($dealer)` instead of a `$carDealer->getCarsForSale()`

Korvin covered everything else I wanted to say I think...

mlocati commented:
> To elaborate on the car example I feel like right now we're essentially doing a Car::getListOfCarsForSale($dealer) instead of a $carDealer->getCarsForSale()

We currently already have `$carDealer->getCarsForSale()`-like methods:
- CollectionKey::getList
- FileCategory::getList
- UserCategory::getList

Since all those getList methods share 99% of the code, they use a common method from their parent class (was named getList, but now it's named getAttributeKeyList).
...traits would simplify all this stuff a lot...
For the moment, I agree with @aembler: let's focus on making concrete5 working with PHP7 while keeping backward compatibility (and let's hope I'll buy my new car, it's really nice :wink: ).

KorvinSzanto commented:
@mlocati I'm fine with whatever you and @aembler decide to do, I just want to make sure you understand the argument that I'm presenting because I believe software design to be very important. It doesn't seem like you responded to ANY of the points that I made, do you mind answering to them?

mlocati commented:
@KorvinSzanto I'd prefer to make concrete5 working with PHP 7, in my limited spare time, but ok, let's see my thoughts and the reasons of what I wrote...

> SRP doesn't speak to static vs non-static, just that classes have only a single reason to change

I wrote that a strict approach to SRP would lead to avoid static methods because you said "This class should be a value object, it shouldn't be doing anything other than being an attribute key according to SRP"
So, this class would have only instance-related methods (eg `$car->sell`) and not static methods (eg `Car::getListOfCarsForSale($dealer)`.


> I think you're looking at this wrong. In real life, I do not buy tell my car what dealership to purchase itself from, in reality the car has almost nothing to do with the purchase.

That's the reason why it's static. But maybe that's a wrong example (the Giulietta is drifting my thoughts these days :wink:). What about something like these methods?
Color::createFromRGB($r, $g, $b);
Color::createFromCMYK($c, $m, $y, $k);

> We have SOLID for a reason @mlocati I think you'd really do yourself a favor by reading into SRP and SOLID as a whole.

I already read something, but obviously not as much as you do.
That's because of the kind of my work and personal interests, which are very diversified.
For instance, last week I had to setup a mail server (postfix, dovecot, sieve, amavis, clamav, roundcube), read and study some old VB6 applications and translate them to PHP, while fixing a C# application that control the shipments of our physical products. In the meanwhile I trained a collegue of mine about how to use a GUI tool that I created to extract some statistical info about our contacts and sales, for which I created an interface with an external service to send huge numbers of emails.
Long short story: you give the direction to concrete5, you define its direction. I fix little aspects of it in my little spare time.

KorvinSzanto commented:
> Long short story: you give the direction to concrete5, you define its direction. I fix little aspects of it in my little spare time.

I don't give concrete5 direction, I'm a member of this community as much as you are. If you don't think that your opinion matters in this case that's your prerogative but I think you have a lot more to bring - and DO bring a lot more - to the table than just little fixes so don't sell yourself too short ;)

> So, this class would have only instance-related methods (eg $car->sell) and not static methods (eg Car::getListOfCarsForSale($dealer).

That's not necessarily true, a value object can still have static constructor methods. SRP would require you have a `Factory` class for returning new instances though.

> Color::createFromRGB($r, $g, $b);
> Color::createFromCMYK($c, $m, $y, $k);

These are constructor methods, they return new instances of static. This does not violate SRP AFAIK, but it may violate OCP as changing how objects are constructed would require you to edit the class in place.

> I already read something, but obviously not as much as you do.

I apologize if offering this link to you was offensive, though if you mean to say there's somehow something wrong with reading about software design I think that's a mistake. I found this link and read through it just last night and learned a lot about SRP and SOLID.

Like I've said a few times before, I'm not opposed to this getting in I just want to get this conversation started.

Thanks,
Korvin

mlocati commented:
@aembler What about accepting (or refusing) this PR before some merge conflict happens?
